demoapp/demo_app.py

```diff
@@ -58,224 +58,430 @@ def get_mgl_sections_file() -> pd.DataFrame:
+                """
 
-    if submit_button:
-        with st.spinner("Working hard..."):
+        prompt = PromptTemplate.from_template(template)
+        
+        chain = create_stuff_documents_chain(llm, prompt)
 
-            category_response = generate_categories(bill_content)
-            response= generate_response(bill_number,bill_content, category_response)
-            #tag_response = generate_tags(category_response, bill_content)
+        with get_openai_callback() as cb:
+            response = chain.invoke({"context": bill_text, "mgl_sections": mgl_documents, "title": bill_title, "names": mgl_names, "categories": category_for_bill, "tags": tags_for_bill})
+            st.write(cb.total_tokens, cb.prompt_tokens, cb.completion_tokens, cb.total_cost)     
+    return response
     
-            with col1:
-                st.subheader(f"Original Bill: #{bill_number}")
-                st.write(bill_title)
-                st.write(bill_content)
-
-            with col2:
-                st.subheader("Generated Text")
-                st.write(response)
-                st.write("###")
-                    
-            with col3:
-                st.subheader("Evaluation Metrics")
-                # rouge score addition
-                scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)
-                rouge_scores = scorer.score(bill_content, response)
-                st.write(f"ROUGE-1 Score: {rouge_scores['rouge1'].fmeasure:.2f}")
-                st.write(f"ROUGE-2 Score: {rouge_scores['rouge2'].fmeasure:.2f}")
-                st.write(f"ROUGE-L Score: {rouge_scores['rougeL'].fmeasure:.2f}")
+def generate_response(df: pd.DataFrame, bill_number: str, bill_title: str, bill_text: str, mgl_ref: str) -> str:
+    """
+    Determines document size and generates an appropriate response for the bill.
+    
+    Args:
+        df (pd.DataFrame): DataFrame containing bill information.
+        bill_number (str): Number of the bill.
+        bill_title (str): Title of the bill.
+        bill_text (str): Text content of the bill.
+    
+    Returns:
+        str: Generated response for the bill.
+    """
+    API_KEY = st.session_state["OPENAI_API_KEY"]
+    os.environ['OPENAI_API_KEY'] = API_KEY
+    llm = ChatOpenAI(openai_api_key=API_KEY, temperature=0, model='gpt-4-1106-preview', model_kwargs={'seed': 42})  
+    
+    mgl_names = get_chap_sec_names(df, bill_number)
+    
+    #Context size for 'gpt-4-1106-preview' is 128K; Need to break up the documents and use vector embenddings when the word length of the documents is large
+    #Chose 90K cut off for large docs because: 1 token = 0.75 words --> 90000 words = ~120K words + leaving  room for response
+    if bill_text is None or str(bill_text).strip().lower() =="nan" or bill_text == "":
+        response = f"Bill {bill_number} is not available for summarization"
+    
+    elif (len(mgl_ref.split()) + len (bill_text.split())) < 90000:
```

Comment:
It would be better to actually calculate the number of tokens of the filled template. I think langchain has some utilities for doing that. But I bet this works reasonably well.

Reply:
got it. Using tiktoken to compute the number of tokens.